### ISSUE TITLE
Regression: modifier keys on physical keyboard switch layers on touch keyboard

### ISSUE THREAD

**Issue by mcdurdin** [keymanapp/keyman#311]
In KeymanWeb 2.0.473, when using a physical keyboard in combination with the emulated Keyman touch keyboard (e.g. in the Chrome mobile emulator), the physical keyboard had no effect: keystrokes would generate no output. This was by design.

In alpha 10.0.15, keystrokes on the physical keyboard do have effects. This is not necessarily a bad thing in theory, but the keystrokes are not flowing through the correct 'desktop' pathways, meaning that keyboards have incorrect behaviours.

So, somewhere we have a regression in how physical keystrokes are processed in mobile emulation mode.

Ideal resolution:

1. Physical keystrokes are processed according to desktop semantics (including applying correct platform values) but touch keystrokes continue to use the touch pathway.
2. Modifier and State keys have no impact on the touch keyboard. *It is possible to make standard modifiers reflect on the touch keyboard, but the layout will not necessarily match the physical keyboard anyway so it's more confusing than not; plus there is no physical 'numeric' modifier key so it is a faulty paradigm anyway!)

----

Conversation here recorded for posterity:

marc [28 minutes ago] 
@jahorton I'm seeing some weird behaviour when simulating the touch keyboard now -- pressing modifier keys on the physical keyboard changes the layer for the touch keyboard, which it didn't used to.

I don't think it is directly related to the chirality PR, but the issue is present in alpha 10.0.15 and not in release 2.0.473.

In Chrome with mobile emulation on:
* keymanweb.com?tier=alpha shows the problem
* keymanweb.com shows the correct behaviour.

jahorton [21 minutes ago] 
Um... keymanweb.com doesn't exactly have the correct behavior for me.  It doesn't take _any_ physical key input.  I feel like this is something I corrected for in a relatively recent PR, so I may want to revisit that to ensure I didn't go too far with its changes.

jahorton [19 minutes ago] 
It can't be related to chirality if it's in the live alpha.  I wonder how it works with the new PR I put up, https://github.com/keymanapp/keyman/pull/310.

At the least, that PR is addressing a highly-related issue.
GitHub
[Web] Next layer processing fix by jahorton · Pull Request #310 · keymanapp/keyman
Addresses #116. Turns out that after #117's work toward #115, little extra work actually needed to be done to correct #116, though this took some time and investigation to determine.
 

jahorton [18 minutes ago] 
I almost broke the desktop OSK with my initial change, which I then had to unbreak with the last commit of that PR.  That 'extraneous statement' may be at fault for the present issue, but will be filtered to 'desktop'-only.

marc [11 minutes ago] 
The previous behaviour on desktop was that in emulation mode it would not accept any physical keystroke. There's some work we can do there but there's more to it than just accepting the keystrokes (c.f. Android hardware keyboard support).

jahorton [4 minutes ago] 
Android hardware keyboard support does some significant manipulation of the OSK that basically cuts the wires, if I remember correctly.  @darcywong00, you haven't seen the OSK visually respond to any physical keystrokes while working on chirality, right?

darcy [3 minutes ago] 
iirc, the OSK stays at the current layer, and the app inserts the correct character for your key combo

marc [2 minutes ago] 
Correct. That same string-cutting needs to be done -- i.e. platform should return as desktop not mobile/tablet and therefore desktop rules should fire (and osk manipulation should not take place). I don't think we try and solve that this afternoon because I'd like a bigger discussion before we do it

**Comment by jahorton** [comment]
As of #326, this has been reverted to the prior "disabled" mode since further investigations and design work are needed for proper implementation.  Hooks have been left in the code to minimize the known work needed when the time is right, but this issue will be parked at this stage for now.
